### PR TITLE
planner: support cross shard DELETE with LIMIT/ORDER BY

### PR DESCRIPTION
**Pull request by systay**
## Description

This PR add support for `Delete` with `Limit` optionally having `Order By` for sharded keyspaces.

## Related Issue(s)
- https://github.com/vitessio/vitess/issues/6854

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [ ] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation was added or is not required

### CODE REVIEW

File: go/vt/vtgate/planbuilder/operators/delete.go [hunk 1]
```diff
@@ -78,12 +76,6 @@ func (d *Delete) GetOrdering(*plancontext.PlanningContext) []OrderBy {
...
 	limit := ""
 	orderBy := ""
-	if d.Limit != nil {
-		limit = " " + sqlparser.String(d.Limit)
-	}
-	if len(d.OrderBy) > 0 {
-		orderBy = " " + sqlparser.String(d.OrderBy)
-	}
 
 	return fmt.Sprintf("%s.%s%s%s", d.Target.VTable.Keyspace.Name, d.Target.VTable.Name.String(), orderBy, limit)
```
Comment (harshit-gangal) [inline go/vt/vtgate/planbuilder/operators/delete.go:80]: nit: printing of limit and order by should be removed completely. printing empty strings.

File: go/vt/vtgate/planbuilder/operators/query_planning.go [hunk 2]
```diff
@@ -108,21 +108,15 @@ func runRewriters(ctx *plancontext.PlanningContext, root Operator) Operator {
...
-func tryPushDelete(ctx *plancontext.PlanningContext, in *Delete) (Operator, *ApplyResult) {
+func tryPushDelete(in *Delete) (Operator, *ApplyResult) {
 	switch src := in.Source.(type) {
 	case *Route:
 		return pushDeleteUnderRoute(in, src)
-	case *ApplyJoin:
-		return pushDeleteUnderJoin(ctx, in, src)
 	}
 	return in, nil
 }
```
Comment (harshit-gangal) [inline go/vt/vtgate/planbuilder/operators/query_planning.go:117]: nit: can be changes to a type check than a switch.

File: go/vt/vtgate/engine/delete_with_input.go [hunk 3]
```diff
@@ -0,0 +1,121 @@
+/*
+Copyright 2023 The Vitess Authors.
```
Comment (frouioui) [inline go/vt/vtgate/engine/delete_with_input.go:2]:
```suggestion
Copyright 2024 The Vitess Authors.
```

File: go/vt/vtgate/engine/delete_with_input_test.go [hunk 4]
```diff
@@ -0,0 +1,116 @@
+/*
+Copyright 2023 The Vitess Authors.
```
Comment (frouioui) [inline go/vt/vtgate/engine/delete_with_input_test.go:2]:
```suggestion
Copyright 2024 The Vitess Authors.
```

File: go/vt/vtgate/planbuilder/operators/delete_with_input.go [hunk 5]
```diff
@@ -0,0 +1,83 @@
+/*
+Copyright 2023 The Vitess Authors.
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/operators/delete_with_input.go:2]:
```suggestion
Copyright 2024 The Vitess Authors.
```

File: go/test/endtoend/vtgate/queries/dml/dml_test.go [hunk 6]
```diff
@@ -78,3 +79,64 @@ func TestMultiTableDelete(t *testing.T) {
 	mcmp.AssertMatches(`select oid, ename from oevent_tbl order by oid`,
 		`[[INT64(1) VARCHAR("a")] [INT64(2) VARCHAR("b")] [INT64(3) VARCHAR("a")] [INT64(4) VARCHAR("c")]]`)
 }
+
+// TestDeleteWithLimit executed delete queries with limit
+func TestDeleteWithLimit(t *testing.T) {
+	utils.SkipIfBinaryIsBelowVersion(t, 19, "vtgate")
+
```
Comment (frouioui) [inline go/test/endtoend/vtgate/queries/dml/dml_test.go:86]: This is a good test case, it looks good to me

File: go/vt/vtgate/engine/delete_with_input.go [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+	topodatapb "vitess.io/vitess/go/vt/proto/topodata"
+	"vitess.io/vitess/go/vt/vterrors"
+
+	"vitess.io/vitess/go/sqltypes"
+	querypb "vitess.io/vitess/go/vt/proto/query"
+)
+
+var _ Primitive = (*DeleteWithInput)(nil)
+
+const DM_VALS = "dm_vals"
```
Comment (frouioui) [inline go/vt/vtgate/engine/delete_with_input.go:31]: nit: I get a warning here that it does not follow the snake case

**Review by frouioui — APPROVED** [review]

**Review by GuptaManan100 — APPROVED** [review]
LGTM!

**Pull request merged**